### PR TITLE
Batch migrate

### PR DESCRIPTION
**Pull request by sanandnarayan**
# Batch Migration

## Description

This PR adds a new contract called CrabMigration which allows for migrating users from v1 to v2 in a batch. 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document update

## How Has This Been Tested

The proper way to test this would be to run the test/e2e/crabMigration.ts test on mainnet fork. The unit test needs too much mocking and isn't a good way to test the flow. 

## User Facing Checklist

- [x] I fully understand the user problem this PR is solving
- [X] I know who the target user is for this PR and have a deep understanding of that user
- [X] I have tried this flow thinking from the pov of the target user for this PR
- [ ] I (or working w someone on team) have scheduled a user test for this PR (if it is a large change)

### CODE REVIEW

File: packages/hardhat/contracts/mocks/MockEuler.sol [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity =0.7.6;
+pragma abicoder v2;
+
+
+contract MockEuler {
+    uint256 a; 
+    function deferLiquidityCheck(address account, bytes memory data) external { 
+        a = 4;
```
Comment (alpinechicken) [inline packages/hardhat/contracts/mocks/MockEuler.sol:10]: why 4?

File: packages/hardhat/contracts/strategy/CrabStrategyV2.sol [hunk 2]
```diff
@@ -181,11 +188,44 @@ contract CrabStrategyV2 is StrategyBase, StrategyFlashSwap, ReentrancyGuard, Own
         require(msg.sender == weth || msg.sender == address(powerTokenController), "Cannot receive eth");
     }
 
+    /**
+     * @notice initializes the collateral ratio upon the first migration
+     */
+     function initialize(uint256 wSqueethToMint) external payable {
```
Comment (aparnakr) [inline packages/hardhat/contracts/strategy/CrabStrategyV2.sol:194]: note to myself: Remember to only allow CrabMigration to call this if not big problems.

**Review by aleone — CHANGES_REQUESTED** [review]
Need to look at tests more once we decide to do with shares, but giving comments on solidity w/o much look at tests.

File: packages/hardhat/contracts/strategy/CrabMigration.sol [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+     /**
+      * @notice allows users to claim their amount of crab v2 shares
+      */
+     function claimV2Shares() external afterMigration { 
+         uint256 amountV1Deposited = sharesDeposited[msg.sender];
+         sharesDeposited[msg.sender] = 0;
+         uint256 crabV2TotalShares = crabV2.balanceOf(address(this));
+         uint256 amountV2ToTransfer = amountV1Deposited * totalCrabV2SharesReceived / totalCrabV1SharesMigrated;
```
Comment (alpinechicken) [inline packages/hardhat/contracts/strategy/CrabMigration.sol:145]: How does this go with precision? Could end up taking 1 more or less from rounding?

File: packages/hardhat/contracts/strategy/CrabStrategyV2.sol [hunk 4]
```diff
@@ -744,23 +756,9 @@ contract CrabStrategyV2 is StrategyBase, StrategyFlashSwap, ReentrancyGuard, Own
         uint256 wSqueethToMint;
         uint256 feeAdjustment = _calcFeeAdjustment();
 
-        if (_strategyDebtAmount == 0 && _strategyCollateralAmount == 0) {
```
Comment (aleone) [inline packages/hardhat/contracts/strategy/CrabStrategyV2.sol:747]: we need this and next line, see was added back in #547 !

File: packages/hardhat/contracts/strategy/CrabMigration.sol [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+// interface
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import {IEulerExec, IEulerDToken} from "../interfaces/IEuler.sol";
+import {WETH9} from "../external/WETH9.sol";
+import "hardhat/console.sol";
+
+// contract
+import {CrabStrategyV2} from "./CrabStrategyV2.sol";
```
[the remark this answers is not in the document]
Comment (aparnakr) [reply packages/hardhat/contracts/strategy/CrabMigration.sol:14]: there just wasn't a CrabV1 and a CrabV2 interface and with a lot of moving targets was easier to import contract

File: packages/hardhat/contracts/strategy/CrabStrategyV2.sol [hunk 6]
```diff
@@ -181,11 +188,36 @@ contract CrabStrategyV2 is StrategyBase, StrategyFlashSwap, ReentrancyGuard, Own
         require(msg.sender == weth || msg.sender == address(powerTokenController), "Cannot receive eth");
     }
 
+    /**
+     * @notice initializes the collateral ratio upon the first migration
+     */
+    function initialize(uint256 wSqueethToMint, uint256 crabSharesToMint) external payable {
```
Comment (alpinechicken) [inline packages/hardhat/contracts/strategy/CrabStrategyV2.sol:194]: is it intentionally allowing it to initialize again if debt/collateral = 0?

**Review by alpinechicken — APPROVED** [review]
lgtm - is the migration test waiting for some merges?

File: packages/hardhat/contracts/strategy/CrabMigration.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (KMKoushik) [reply packages/hardhat/contracts/strategy/CrabMigration.sol]: Fixing this in #543

File: packages/hardhat/contracts/strategy/CrabMigration.sol [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    /** 
+     * @notice migration constructor
+     * @param _crabV1 address of crab v1
+     * @param _crabV2 address of crab v2 
+     */
+     constructor (address payable _crabV1, address payable _crabV2, address _weth, address _eulerExec, address _dToken, address _eulerMainnet) { 
+         crabV1 = CrabStrategy(_crabV1);
+         crabV2 = CrabStrategyV2 (_crabV2);
```
[the remark this answers is not in the document]
Comment (aparnakr) [reply packages/hardhat/contracts/strategy/CrabMigration.sol]: So we should know the address of crabV2 for migration. Was thinking in CrabV2 there will be a method to set the migrationAddress before initialization.

**Comment by aparnakr** [comment]
> lgtm - is the migration test waiting for some merges?

it should be good to review!

File: packages/hardhat/contracts/strategy/CrabStrategyV2.sol [hunk 8]
```diff
@@ -134,6 +136,11 @@ contract CrabStrategyV2 is StrategyBase, StrategyFlashSwap, ReentrancyGuard, Own
         _;
     }
 
+    modifier afterInitialization() {
```
[the remark this answers is not in the document]
Comment (aparnakr) [reply packages/hardhat/contracts/strategy/CrabStrategyV2.sol:140]: yes this makes sense!

File: packages/hardhat/contracts/strategy/CrabStrategyV2.sol [hunk 9]
```diff
@@ -181,11 +192,44 @@ contract CrabStrategyV2 is StrategyBase, StrategyFlashSwap, ReentrancyGuard, Own
...
+        uint256 strategyDebt;
+        uint256 strategyCollateral;
+
+        _checkStrategyCap(amount, strategyCollateral);
+
+        require((strategyDebt == 0 && strategyCollateral == 0), "C5");
+        // store hedge data as strategy is delta neutral at this point
+        // only execute this upon first deposit
+        uint256 wSqueethEthPrice = IOracle(oracle).getTwap(ethWSqueethPool, wPowerPerp, weth, hedgingTwapPeriod, true);
+        timeAtLastHedge = block.timestamp;
```
Comment (alpinechicken) [inline packages/hardhat/contracts/strategy/CrabStrategyV2.sol:217]: timeAtLastHedge and priceAtLastHedge shadowed inputs

Comment (aparnakr) [reply packages/hardhat/contracts/strategy/CrabStrategyV2.sol:217]: What do you mean shadow inputs?

Comment (alpinechicken) [reply packages/hardhat/contracts/strategy/CrabStrategyV2.sol:217]: they're coming in as params on 201 and 202 then gettting set on 217 and 218 so the inputs don't do anything

File: packages/hardhat/test/e2e/crab-Migration.ts [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+            oracleAddress,
+            wethAddress,
+            uniswapFactoryAddress,
+            wethOsqthPoolAddress,
+            timelock.address,
+            crabMigration.address,
+            1,
+            1)) as CrabStrategyV2;
+        await crabStrategyV2.setStrategyCap(ethers.utils.parseEther("1000.0"));
+        await crabMigration.connect(owner).setCrabV2(crabStrategyV2.address);
```
Comment (alpinechicken) [inline packages/hardhat/test/e2e/crab-Migration.ts:103]: what dis?

File: packages/hardhat/test/e2e/crab-Migration.ts [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+            const d2EthBalanceBefore = await provider.getBalance(d2.address);
+
+            const [ethToGet, sqthToSell] = await getV2SqthAndEth(d2sharesV2ToMigrate);
+            const sqthPrice = await oracle.getTwap(wethOsqthPoolAddress, squeethAddress, wethAddress, 1, false)
+            const ethSpentToBuyBack = wdiv(wmul(wmul(sqthToSell, sqthPrice), BigNumber.from(105)), BigNumber.from(100))
+
+
+            await crabMigration.connect(d2).claimAndWithdraw(d2sharesV2ToMigrate, ethSpentToBuyBack) // Set ETH slippage higher!
+            const d2sharesInMigrationAfter = await crabMigration.sharesDeposited(d2.address)
+            const constractSharesAfter = await crabStrategyV2.balanceOf(crabMigration.address);
```
Comment (alpinechicken) [inline packages/hardhat/test/e2e/crab-Migration.ts:263]: constract lol

File: packages/hardhat/test/unit-tests/strategy/crab-migration.ts [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+        it("random should not be able to call onDeferredLiquidity()", async () => { 
+            const data = "0x0000000000000000000000000000000000000000"
+            await expect((crabMigration.connect(random).onDeferredLiquidityCheck(data))).to.be.revertedWith("M4");
+        })
+
+        it("random should not be able to migrate shares", async () => { 
+            await expect(crabMigration.connect(random).batchMigrate()).to.be.revertedWith("Ownable: caller is not the owner");
+        })
+
+        it("batchMigrate", async () => {
```
Comment (sanandnarayan) [inline packages/hardhat/test/unit-tests/strategy/crab-migration.ts:129]: should we remove this case? it doesnt do any test

**Comment by sanandnarayan** [comment]
@KMKoushik can you approve this on my behalf please . Thank you

**Review by KMKoushik — APPROVED** [review]

**Review by aparnakr — APPROVED** [review]

**Pull request merged**